RecoLocalTracker/SiPhase2VectorHitBuilder/src/VectorHitBuilderAlgorithm.cc

```diff
@@ -0,0 +1,553 @@
+    }
+
+    const CurvatureAndPhi curvatureAndPhi = curvatureANDphi(gPositionLower, gPositionUpper, gErrorLower, gErrorUpper);
+    VectorHit vh = VectorHit(*stack,
+                             vh2Dzx,
+                             vh2Dzy,
+                             lowerOmni,
+                             upperOmni,
+                             curvatureAndPhi.curvature,
+                             curvatureAndPhi.curvatureError,
+                             curvatureAndPhi.phi);
+    return vh;
+  }
+
+  return VectorHit();
+}
+
+void VectorHitBuilderAlgorithm::fit2Dzx(const Local3DPoint lpCI,
+                                        const Local3DPoint lpCO,
+                                        const LocalError leCI,
+                                        const LocalError leCO,
+                                        Local3DPoint& pos,
+                                        Local3DVector& dir,
+                                        AlgebraicSymMatrix22& covMatrix,
+                                        double& chi2) const {
+  float x[2] = {lpCI.z(), lpCO.z()};
+  float y[2] = {lpCI.x(), lpCO.x()};
+  float sqCI = sqrt(leCI.xx());
+  float sqCO = sqrt(leCO.xx());
+  float sigy2[2] = {sqCI * sqCI, sqCO * sqCO};
+
+  fit(x, y, sigy2, pos, dir, covMatrix, chi2);
+
+  return;
+}
+
+void VectorHitBuilderAlgorithm::fit2Dzy(const Local3DPoint lpCI,
+                                        const Local3DPoint lpCO,
+                                        const LocalError leCI,
+                                        const LocalError leCO,
+                                        Local3DPoint& pos,
+                                        Local3DVector& dir,
+                                        AlgebraicSymMatrix22& covMatrix,
+                                        double& chi2) const {
+  float x[2] = {lpCI.z(), lpCO.z()};
+  float y[2] = {lpCI.y(), lpCO.y()};
+  float sqCI = sqrt(leCI.yy());
+  float sqCO = sqrt(leCO.yy());
+  float sigy2[2] = {sqCI * sqCI, sqCO * sqCO};
+
+  fit(x, y, sigy2, pos, dir, covMatrix, chi2);
+
+  return;
+}
+
+void VectorHitBuilderAlgorithm::fit(float x[2],
+                                    float y[2],
+                                    float sigy2[2],
+                                    Local3DPoint& pos,
+                                    Local3DVector& dir,
+                                    AlgebraicSymMatrix22& covMatrix,
+                                    double& chi2) const {
+  float slope = 0.;
+  float intercept = 0.;
+  float covss = 0.;
+  float covii = 0.;
+  float covsi = 0.;
+
+  linearFit(x, y, 2, sigy2, slope, intercept, covss, covii, covsi);
+
+  covMatrix[0][0] = covss;  // this is var(dy/dz)
+  covMatrix[1][1] = covii;  // this is var(y)
+  covMatrix[1][0] = covsi;  // this is cov(dy/dz,y)
+
+  for (unsigned int j = 0; j < 2; j++) {
+    const double ypred = intercept + slope * x[j];
+    const double dy = (y[j] - ypred) / sqrt(sigy2[j]);
+    chi2 += dy * dy;
```

Comment:
@ebrondol Given that this is a linear fit to two points, is this chi2 variable really meaningful?

Reply:
Hi @JanFSchulte I am not sure I really looked at the distribution of the chi2 of the VHs at the time. One should check the plot.